src/L1/rollup/ScrollChain.sol

```diff
@@ -436,151 +363,61 @@ contract ScrollChain is OwnableUpgradeable, PausableUpgradeable, IScrollChain {
-            _checkBlobVersionedHash(_blobVersionedHash, _blobDataProof);
-            _publicInputHash = keccak256(
-                abi.encodePacked(
-                    layer2ChainId,
-                    _prevStateRoot,
-                    _postStateRoot,
-                    _withdrawRoot,
-                    _dataHash,
-                    _blobDataProof[0:64],
-                    _blobVersionedHash
-                )
-            );
-        }
-
-        // load version from batch header, it is always the first byte.
-        uint256 batchVersion = BatchHeaderV0Codec.getVersion(batchPtr);
-        // verify batch
-        IRollupVerifier(verifier).verifyAggregateProof(batchVersion, _batchIndex, _aggrProof, _publicInputHash);
-
-        // Pop finalized and non-skipped message from L1MessageQueue.
-        uint256 _totalL1MessagesPoppedOverall = BatchHeaderV0Codec.getTotalL1MessagePopped(batchPtr);
-        _popL1MessagesMemory(
-            BatchHeaderV1Codec.getSkippedBitmapPtr(batchPtr),
-            _totalL1MessagesPoppedOverall,
-            BatchHeaderV0Codec.getL1MessagePopped(batchPtr)
-        );
-
-        _afterFinalizeBatch(_totalL1MessagesPoppedOverall, _batchIndex, _batchHash, _postStateRoot, _withdrawRoot);
+    ) external override OnlyProver whenNotPaused whenEnforcedBatchNotEnable {
+        _finalizeBundle(_batchHeader, _postStateRoot, _withdrawRoot, _aggrProof);
     }
 
     /// @inheritdoc IScrollChain
-    function finalizeBundleWithProof(
-        bytes calldata _batchHeader,
-        bytes32 _postStateRoot,
-        bytes32 _withdrawRoot,
-        bytes calldata _aggrProof
-    ) external override OnlyProver whenNotPaused {
-        if (_postStateRoot == bytes32(0)) revert ErrorStateRootIsZero();
-
-        // compute pending batch hash and verify
-        (
-            uint256 batchPtr,
-            bytes32 _batchHash,
-            uint256 _batchIndex,
-            uint256 _totalL1MessagesPoppedOverall
-        ) = _loadBatchHeader(_batchHeader);
-
-        // retrieve finalized state root and batch hash from storage to construct the public input
-        uint256 _finalizedBatchIndex = lastFinalizedBatchIndex;
-        if (_batchIndex <= _finalizedBatchIndex) revert ErrorBatchIsAlreadyVerified();
+    function commitAndFinalizeBatch(CommitStruct calldata commitStruct, FinalizeStruct calldata finalizeStruct)
+        external
+    {
+        // we use BatchHeaderV3Codec for finalizeStruct.batchHeader
+        EnforcedBatchParameters memory parameters = enforcedBatchParameters;
+        if (!parameters.enforcedModeEnabled) {
+            if (parameters.lastFinalizeTimestamp + parameters.maxDelayEnterEnforcedMode < block.timestamp) {
+                // explicit set enforce batch enable
+                parameters.enforcedModeEnabled = true;
+                // reset `lastCommittedBatchIndex`
+                parameters.lastCommittedBatchIndex = uint64(lastFinalizedBatchIndex);
```

Comment:
I think we've discussed this before but I forgot the conclusion; doesn't this make "fast L2 finality" impossible? @jonastheis

In the future with the infallible chain derivation, we could just have a permissionless fallback for batch commitment and finalization (but not need to revert unfinalized batches).

Reply:
With chain derivation we will need to change the mechanism anyway. So I think it is okay to revert here.